RawSpeed/BitStream.h

```diff
@@ -0,0 +1,163 @@
+/*
+    RawSpeed - RAW file decoder.
+
+    Copyright (C) 2009-2014 Klaus Post
+    Copyright (C) 2017 Axel Waggershauser
+
+    This library is free software; you can redistribute it and/or
+    modify it under the terms of the GNU Lesser General Public
+    License as published by the Free Software Foundation; either
+    version 2 of the License, or (at your option) any later version.
+
+    This library is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+    Lesser General Public License for more details.
+
+    You should have received a copy of the GNU Lesser General Public
+    License along with this library; if not, write to the Free Software
+    Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
+*/
+#pragma once
+
+#include "ByteStream.h"
+
+namespace RawSpeed {
+
+// simple 64-bit wide cache implementation that acts like a FiFo.
+// There are two variants:
+//  * L->R: new bits are pushed in on the left and pulled out on the right
+//  * L<-R: new bits are pushed in on the right and pulled out on the left
+// Each BitStream specialization uses one of the two.
+
+struct BitStreamCacheBase
+{
+  uint64 cache = 0; // the acutal bits stored in the cache
+  unsigned int fillLevel = 0; // bits left in cache
+  static constexpr unsigned Size = sizeof(cache)*8;
+  static constexpr unsigned MaxGetBits = Size/2;
+};
+
+struct BitStreamCacheLeftInRightOut : BitStreamCacheBase
+{
+  inline void push(uint64 bits, uint32 count) noexcept {
+    assert(count + fillLevel <= Size);
+    cache |= bits << fillLevel;
+    fillLevel += count;
+  }
+
+  inline uint32 peek(uint32 count) const noexcept {
+    return cache & ((1 << count) - 1);
+  }
+
+  inline void skip(uint32 count) noexcept {
+    cache >>= count;
+    fillLevel -= count;
+  }
+};
+
+struct BitStreamCacheRightInLeftOut : BitStreamCacheBase
+{
+  inline void push(uint64 bits, uint32 count) noexcept {
+    assert(count + fillLevel <= Size);
+    cache = cache << count | bits;
+    fillLevel += count;
+  }
+
+  inline uint32 peek(uint32 count) const noexcept {
+    return (cache >> (fillLevel - count)) & ((1 << count) - 1);
+  }
+
+  inline void skip(uint32 count) noexcept {
+    fillLevel -= count;
+  }
+};
+
+// Note: Allocated buffer MUST be at least FILEMAP_MARGIN bytes larger than
```

Comment:
I would say: "it depends". I think what you mean is that you don't want to have something like this to allow the library to crash if you can accidentally forget about this requirement. Or that this prevents you from implementing some decent fuzzing. Those pitfalls have to go. But this does not generally rule out the idea of internally allocating some more bytes than the file actually has for speed improvements.

I was testing the code with that constant set to 0 and most of the decoders actually worked but some failed (IIRC). So I let it in for now to be addressed later.